test/suite/hook_rollback.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python
+#
+# Public Domain 2014-present MongoDB, Inc.
+# Public Domain 2008-2014 WiredTiger, Inc.
+#
+# This is free and unencumbered software released into the public domain.
+#
+# Anyone is free to copy, modify, publish, use, compile, sell, or
+# distribute this software, either in source code form or as a compiled
+# binary, for any purpose, commercial or non-commercial, and by any
+# means.
+#
+# In jurisdictions that recognize copyright laws, the author or authors
+# of this software dedicate any and all copyright interest in the
+# software to the public domain. We make this dedication for the benefit
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# [TEST_TAGS]
+# ignored_file
+# [END_TAGS]
+
+# hook_rollback.py
+#   Intentionally have random failures of rollback some fraction of the time:
+#     python run.py --hook rollback=FAILRATE base01
+#
+# FAILRATE should be a floating number less than one, for example, 0.25 to fail a quarter of the time.
+#
+# This hook is used to test a feature of the test suite, where tests with rollback errors are automatically
+# restarted.
+
+from __future__ import print_function
+
+import os, sys, wthooks, random, wiredtiger
+from wttest import WiredTigerTestCase
+
+# Print to /dev/tty for debugging, since anything extraneous to stdout/stderr will
+# cause a test error.
+def tty(s):
+    WiredTigerTestCase.tty(s)
+
+# These are the hook functions that are run when particular APIs are called.
+
+def session_begin_transaction_notify(ret, session, *args):
+    session.in_transaction = True
+    #tty('==> in txn, session={}'.format(session))
+
+def session_end_transaction_notify(ret, session, *args):
+    session.in_transaction = False
+    #tty('==> out txn, session={}'.format(session))
+
+# Here we patch a deficiency of our SWIG setup.  When a cursor is created via
+# a session.open_cursor() call, a unique Cursor python object is created.
+# However, that particular 'cursor' object doesn't know that 'session' created it,
+# only that there is a property 'session' that maps to the pointer WT_CURSOR->session
```

Comment:
"the pointer WT_CURSOR->session pointer"  

Cut the first "pointer"?

Reply:
Fixed